### ISSUE TITLE
Documentation or How-to needed

### ISSUE THREAD

**Issue by ralphbecket** [vdom-benchmark/vdom-benchmark#20]
Hi, I'd like to use this benchmark, but I can't find a document or comment anywhere explaining what a client vDOM library needs to provide to participate.  Clearly a bunch of other devs have solved this one, but I'd really appreciate some pointers here, please.

**Comment by localvoid** [comment]
Just fork react implementation https://github.com/vdom-benchmark/vdom-benchmark-react and modify it, so it will use your library. The source code is really simple https://github.com/vdom-benchmark/vdom-benchmark-react/blob/master/web/js/main.jsx

**Comment by leeoniya** [comment]
hey @localvoid 

I gotta say, I'm finding it a bit off-putting and odd that a DOM benchmark not only requires node/npm but also *38MB* of dependencies. I just wanna run a bench of my 10k vdom lib in a browser (which is still required!).

I wrote the needed `main.js` for `domvm` and tried getting the whole thing to run, but then it complained about missing `gulp` (presumably it now needs that in some global path), *ugh*. `domvm` is not on npm yet, since it won't be stable till 4/1, but everything should run fine against `/dist/domvm.min.js`. If you can get it hooked up and running, please do [1], otherwise just ignore my rant :)

thanks!
Leon

[1] https://github.com/leeoniya/domvm/blob/master/test/bench/vdom-benchmark/main.js

**Comment by localvoid** [comment]
>I gotta say, I'm finding it a bit off-putting and odd that a DOM benchmark not only requires node/npm but also 38MB of dependencies.

It only requires https://github.com/vdom-benchmark/vdom-benchmark-base ,everything else is optional. And since most of the libraries require compilation step to build it for production, there are build tools in all repos.

>`domvm` is not on npm yet, since it won't be stable till 4/1

Use `npm link`

**Comment by leeoniya** [comment]
Is there a How-to for running this benchmark having only your `vdom-benchmark-base`, `main.js` and the vdom lib? It is not clear because every bench uses node/npm/gulp.

**Comment by localvoid** [comment]
No, I thought that js developers won't have any problems with the most common tools used for building js projects. Is it so hard just to use node/npm/gulp?

**Comment by leeoniya** [comment]
My viewpoint is, libs that are meant to run in the browser, especially browser-first libs, should provide pre-built `dist` files that are ready to be dropped into a browser and "just work".

Requiring users to first install 75MB of build tools which include a server, a package manager and build system. Then asking them to compile, start the server they just installed and finally open a browser to view a page which serves up html & js from their filesysem is insane. Not only this, but to *use* the lib, also requires them to use the same tools - even if they are "standard".

Take a look at https://github.com/mathieuancelin/js-repaint-perfs. It was trivial to submit a test case. no node, no gulp, no npm. Write js/html/css, make it run, pull request.

domvm's "build" process [1] is concat & min using whatever tools you want. In fact, you can just include all the needed modules in sequential script tags and it will "just work". Cause it's a lib for the browser first, and uses npm only as a distribution/publishing mechanism since domvm is also isomorphic.

There's an embrace of JSX, Babel, and mountains of tooling to write front-end software that just cannot run in a browser without them. Certainly, there are benefits to some of it, TypeScript, ES6-7 cross-compilation, but the whole ecosystem is drunk on tooling. npm, AMD & Browserify should be all that's needed (maybe Babel if you just can't wait for the future). Obviously, not everyone thinks this way :)

Anyways, it's your project, do as you see fit. I don't want to bike-shed.

kivi is great, btw.

[1] https://github.com/leeoniya/domvm#modules-building

**Comment by ralphbecket** [comment]
Hi, thanks for the pointer.  I'm with @leeoniya here: a few words of advice would go a long way.  I couldn't even find vdom-benchmark-base until you provided the link. I'm still unclear on where the starting point is and what is the API I should be working against.

**Comment by localvoid** [comment]
>My viewpoint is, libs that are meant to run in the browser, especially browser-first libs, should provide pre-built dist files that are ready to be dropped into a browser and "just work".

Maybe several years ago it was true, but now javascript community finally learned how to build software.

>Take a look at https://github.com/mathieuancelin/js-repaint-perfs. It was trivial to submit a test case. no node, no gulp, no npm. Write js/html/css, make it run, pull request.

And that is why implementations like React aren't properly compiled in this repo. What is the point to test something, if you aren't doing it in a proper way. Almost all real React projects are built with a set of tools that rewrite/optimize code and benchmark implementation shouldn't be different.

>domvm's "build" process [1] is concat & min using whatever tools you want. In fact, you can just include all the needed modules in sequential script tags and it will "just work". Cause it's a lib for the browser first, and uses npm only as a distribution/publishing mechanism since domvm is also isomorphic.

If you want your library to get widespread adoption, it should work as npm package with browserify/webpack/etc to build final project.

**Comment by localvoid** [comment]
@ralphbecket 
- fork https://github.com/vdom-benchmark/vdom-benchmark-react/
- replace https://github.com/vdom-benchmark/vdom-benchmark-react/blob/master/package.json react dependencies with your package, remove unnecessary dependencies like `"babel-preset-react": "^6.3.13"`
- install all dependencies `npm install`
- change build task https://github.com/vdom-benchmark/vdom-benchmark-react/blob/master/gulpfile.js in a way that you expect your library should be used, maybe replace it with browserify, or something else, simplest way is just to properly write your library in a way that it will work with webpack/browserify/etc.
- replace react api calls with similar api calls in your library https://github.com/vdom-benchmark/vdom-benchmark-react/blob/master/web/js/main.jsx (lines: 5, 6, 14, 16, 33, 37, 41)
- and run `gulp deploy` to deploy your implementation to github pages

**Comment by leeoniya** [comment]
> If you want your library to get widespread adoption, it should work as npm package with browserify/webpack/etc to build final project.

And it will. But it will not _require_ this process from consumers to run in their browser, which is the purpose of any DOM lib. At the very minimum, there will always be a `dist` that anyone can simply drop in and use directly.

I do not believe that projects designed for use in the browser "learned how to build software" by failing to provide releases/dist and instead forcing end users to build them. Javascript is not machine code, it will not compile differently on my machine than on yours, we have exactly the same target architecture/env - there is no reason to force me to install a toolchain to build it. Again, IMO.

**Comment by ralphbecket** [comment]
@localvoid I really do appreciate your response and will give it a go whenever I can find the time: that looks like a substantial amount of learning/installation/tooling for a simple set of benchmarks.
